src/client.rs

```diff
@@ -1,6 +1,288 @@
 pub mod io;
 pub mod messages;
 
-fn main() {
-    println!("Hello, world!");
+use crate::io::router::{RouterBuilder, RouterHandler};
+use anyhow::Result;
+use messages::{
+    requests::{
+        announce_shard_request::AnnounceShardRequest,
+        get_client_shard_info_request::GetClientShardInfoRequest,
+        get_shared_peers_request::GetSharedPeersRequest, get_version_request::GetVersionRequest,
+        query_version_request::QueryVersionRequest, read_request::ReadRequest,
+        write_request::WriteRequest,
+    },
+    responses::{
+        announce_shard_response::AnnounceShardResponse,
+        get_client_shard_info_response::GetClientShardInfoResponse,
+        get_shared_peers_response::GetSharedPeersResponse,
+        get_version_response::GetVersionResponse, query_version_response::QueryVersionResponse,
+        read_response::ReadResponse, write_response::WriteResponse,
+    },
+};
+use std::io::Write;
+
+use std::sync::{Arc, Mutex};
+
+#[derive(Debug, Default, Clone)]
+struct ClientState {
+    num_write_shards: usize,
+    write_shard_info: Vec<(String, u16)>,
+    read_shard_info: Vec<(String, u16)>,
+}
+
+#[derive(Debug, Clone)]
+struct Client {
+    server_address: String,
+    shard_state: Arc<Mutex<ClientState>>,
+}
+
+impl Client {
+    fn new(server_address: String, shard_state: Arc<Mutex<ClientState>>) -> Self {
+        Client {
+            server_address,
+            shard_state,
+        }
+    }
+}
+
+impl RouterHandler for Client {
+    fn handle_write_request(&self, _req: &WriteRequest) -> WriteResponse {
+        unimplemented!()
+    }
+
+    fn handle_read_request(&self, _req: &ReadRequest) -> ReadResponse {
+        unimplemented!()
+    }
+
+    fn handle_get_client_shard_info_response(&self, res: &GetClientShardInfoResponse) {
+        println!("Received shard information from main info server:");
+
+        let num_write_shards = res.num_write_shards as usize;
+        let write_shard_info: Vec<(String, u16)> = res
+            .write_shard_info
+            .iter()
+            .map(|(ip, port)| (format!("{}.{}.{}.{}", ip[0], ip[1], ip[2], ip[3]), *port))
+            .collect();
+
+        let read_shard_info: Vec<(String, u16)> = res
+            .read_shard_info
+            .iter()
+            .map(|(ip, port)| (format!("{}.{}.{}.{}", ip[0], ip[1], ip[2], ip[3]), *port))
```

Comment:
I think are using ipv6 so this seems kind of wrong. I would initialize a u128 number from the entire byte sequence (something like u128::from_le_something) and then use the from from_bits on https://doc.rust-lang.org/std/net/struct.Ipv6Addr.html. It might also makes sense to change the router to use Ipv6Addr structs instead of the strings so we don't have to keep converting back and forth between strings

Reply:
I ended up changing the router to use SocketAddrIpv6 so look through the recent changes to make this work